### PR TITLE
Implement file cache for s3 filesystem

### PR DESCRIPTION
**Pull request by acquamarin**
# Description

This PR introduces the file cache for s3 filesystem. Removes the default clientContext from the virtualFileSystem. Now, callers of the FileSystem::openFile() needs to pass in a clientContext for accessing the client configuration purpose.

# Contributor agreement

- [x] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

File: extension/httpfs/src/httpfs.cpp [hunk 1]
```diff
@@ -113,8 +113,8 @@ void HTTPFileInfo::initialize() {
             // LCOV_EXCL_STOP
         }
     }
-    if (httpConfig.cacheFile) {
-        cachedFileInfo = hfs->getCachedFileManager().getCachedFileInfo(path);
+    if (httpConfig.cacheFile && length > CachedFileManager::MAX_SEGMENT_SIZE * 5) {
```
Comment (ray6080) [inline extension/httpfs/src/httpfs.cpp:116]: Should we still do this optimization 😅 ? `length > CachedFileManager::MAX_SEGMENT_SIZE * 5`

File: extension/httpfs/src/cached_file_manager.cpp [hunk 2]
```diff
@@ -45,13 +47,16 @@ std::string CachedFileManager::getCachedFilePath(const std::string& originalFile
...
 }
 
-void CachedFileManager::downloadFile(const std::string& path, FileInfo* info) {
-    auto url = HTTPFileSystem::parseUrl(path);
-    httplib::Client cli(url.first);
-    httplib::Headers headers = {{}};
-    auto fileContent = cli.Get(url.second, headers)->body;
-    info->writeFile(reinterpret_cast<const uint8_t*>(fileContent.c_str()), fileContent.size(),
-        0 /* offset */);
+void CachedFileManager::downloadFile(HTTPFileInfo* fileToDownload, FileInfo* info) {
```
Comment (ray6080) [inline extension/httpfs/src/cached_file_manager.cpp:50]: info -> cacheFileInfo / localFileInfo.

File: extension/httpfs/src/cached_file_manager.cpp [hunk 3]
```diff
@@ -45,13 +47,16 @@ std::string CachedFileManager::getCachedFilePath(const std::string& originalFile
...
+    uint64_t numBytesRead;
+    uint64_t offsetToWrite = 0;
+    do {
+        numBytesRead = fileToDownload->readFile(downloadBuffer.get(), MAX_SEGMENT_SIZE);
+        info->writeFile(downloadBuffer.get(), numBytesRead, offsetToWrite);
+        offsetToWrite += numBytesRead;
+    } while (numBytesRead != 0);
```
Comment (ray6080) [inline extension/httpfs/src/cached_file_manager.cpp:57]: Let's do some benchmark here. I'm not sure about the performance implication here, but shouldn't a bit larger buffer size (but still multiple of `MAX_SEGMENT_SIZE`) will yield better throughput from httpfs/s3? Can u give a try on larger sizes?

File: src/storage/file_handle.cpp [hunk 4]
```diff
@@ -37,12 +39,9 @@ void FileHandle::constructExistingFileHandle(const std::string& path, VirtualFil
     }
 }
 
-void FileHandle::constructNewFileHandle(const std::string& path) {
-#ifdef _WIN32
-    fileInfo = make_unique<FileInfo>(path, nullptr /* fs */);
-#else
-    fileInfo = make_unique<FileInfo>(path, nullptr /* fs */);
-#endif
+void FileHandle::constructNewFileHandle(const std::string& path, VirtualFileSystem* vfs,
```
Comment (ray6080) [inline src/storage/file_handle.cpp:42]: Nice catch here 👍

File: src/storage/storage_manager.cpp [hunk 5]
```diff
@@ -18,25 +18,28 @@ namespace storage {
...
     : databasePath{databasePath}, readOnly{readOnly}, memoryManager{memoryManager},
-      enableCompression{enableCompression}, vfs{vfs} {
-    dataFH = initFileHandle(StorageUtils::getDataFName(vfs, databasePath));
-    metadataFH = initFileHandle(StorageUtils::getMetadataFName(vfs, databasePath));
+      enableCompression{enableCompression} {
+    dataFH = initFileHandle(StorageUtils::getDataFName(vfs, databasePath), vfs, context);
+    metadataFH = initFileHandle(StorageUtils::getMetadataFName(vfs, databasePath), vfs, context);
     // TODO: Assert no wal file exists.
-    wal = std::make_unique<WAL>(databasePath, readOnly, *memoryManager.getBufferManager(), vfs);
+    // TODO: should we disable WAL in readonly mode?
```
Comment (ray6080) [inline src/storage/storage_manager.cpp:27]: Eventually I think so.

**Review by ray6080 — APPROVED** [review]

File: extension/httpfs/src/cached_file_manager.cpp (hunk 3 above)
Comment (acquamarin) [reply extension/httpfs/src/cached_file_manager.cpp:57]: You are right, a relatively bigger segment size yields a better performance

File: extension/httpfs/src/httpfs.cpp (hunk 1 above)
Comment (acquamarin) [reply extension/httpfs/src/httpfs.cpp:116]: removed

**Pull request merged**